pkg/hub/addon/discovery_controller.go

```diff
@@ -0,0 +1,246 @@
+	addOnStatusUnreachable = "unreachable"
+)
+
+// addOnFeatureDiscoveryController monitors ManagedCluster and its ManagedClusterAddOns on hub and
+// create/update/delete labels of the ManagedCluster to reflect the status of addons.
+type addOnFeatureDiscoveryController struct {
+	clusterClient clientset.Interface
+	clusterLister clusterv1listers.ManagedClusterLister
+	addOnLister   addonlisterv1alpha1.ManagedClusterAddOnLister
+	recorder      events.Recorder
+}
+
+// NewAddOnFeatureDiscoveryController returns an instance of addOnFeatureDiscoveryController
+func NewAddOnFeatureDiscoveryController(
+	clusterClient clientset.Interface,
+	clusterInformer clusterv1informer.ManagedClusterInformer,
+	addOnInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addOnFeatureDiscoveryController{
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		addOnLister:   addOnInformers.Lister(),
+		recorder:      recorder,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				key, _ := cache.MetaNamespaceKeyFunc(obj)
+				return key
+			},
+			addOnInformers.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(10*time.Minute).
+		ToController("AddOnFeatureDiscoveryController", recorder)
+}
+
+func (c *addOnFeatureDiscoveryController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// The value of queueKey might be
+	// 1) equal to the default queuekey. It is triggered by resync every 10 minutes;
+	// 2) in format: namespace/name. It indicates the event source is a ManagedClusterAddOn;
+	// 3) in format: name. It indicates the event source is a ManagedCluster;
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	switch {
+	case queueKey == factory.DefaultQueueKey:
+		// handle resync
+		return c.resync(syncCtx)
+	case len(namespace) > 0:
+		// sync a particular addon
+		return c.syncAddOn(ctx, namespace, name)
+	default:
+		// sync the cluster
+		return c.syncCluster(ctx, name)
+	}
+}
+
+func (c *addOnFeatureDiscoveryController) resync(syncCtx factory.SyncContext) error {
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, cluster := range clusters {
+		if !cluster.DeletionTimestamp.IsZero() {
+			continue
+		}
+		syncCtx.Queue().Add(cluster.Name)
+	}
+	return nil
+}
+
+func (c *addOnFeatureDiscoveryController) syncAddOn(ctx context.Context, clusterName, addOnName string) error {
+	klog.V(4).Infof("Reconciling addOn %q", addOnName)
+
+	addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
+	if errors.IsNotFound(err) {
```

Comment:
I might get cluster here, merge labels, then update cluster. So we do not need to differentiate deleteAddonLabel and updateAddonLabel.

Reply:
Code updated as suggested.

Reply:
what i  meant is

```go

labels := map[string]string{}
addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
switch {
case errors.IsNotFound(err):
  labels[addonName + "-"] = ""
case err != nil:
  return err
case !addOn.DeletionTimestamp.IsZero():
  labels[addonName + "-"] = ""
default:
  labels[addonName] = getAddOnLabelValue(addOn)
}

cluster, err := c.clusterLister.Get(clusterName)
if err != nil {
	return fmt.Errorf("unable to find cluster with name %q: %w", clusterName, err)
}

resourcemerge.MergeMap(&modified, &cluster.Labels,  labels)

if modified {
 // update things
}

```